src/nvim/move.c

```diff
@@ -1691,27 +2102,68 @@ void scroll_cursor_bot(int min_scroll, int set_topbot)
 ///
 void scroll_cursor_halfway(bool atend, bool prefer_above)
 {
-  int above = 0;
-  int topfill = 0;
-  int below = 0;
-  lineoff_T loff;
-  lineoff_T boff;
   linenr_T old_topline = curwin->w_topline;
-
-  loff.lnum = boff.lnum = curwin->w_cursor.lnum;
+  lineoff_T loff = { .lnum = curwin->w_cursor.lnum };
+  lineoff_T boff = { .lnum = curwin->w_cursor.lnum };
   (void)hasFolding(loff.lnum, &loff.lnum, &boff.lnum);
   int used = plines_win_nofill(curwin, loff.lnum, true);
   loff.fill = 0;
   boff.fill = 0;
   linenr_T topline = loff.lnum;
+  colnr_T skipcol = 0;
+  bool set_skipcol = false;
+
+  int half_height = 0;
+  bool smooth_scroll = false;
+  if (curwin->w_p_sms && curwin->w_p_wrap) {
+    // 'smoothscroll' and 'wrap' are set
+    smooth_scroll = true;
+    half_height = (curwin->w_height - used) / 2;
+    used = 0;
+  }
 
+  int topfill = 0;
   while (topline > 1) {
+    // If using smoothscroll, we can precisely scroll to the
+    // exact point where the cursor is halfway down the screen.
+    if (smooth_scroll) {
+      topline_back_winheight(curwin, &loff, false);
+      if (loff.height == MAXCOL) {
+        break;
+      } else {
+        used += loff.height;
+      }
+      if (used > half_height) {
+        if (used - loff.height < half_height) {
+          int plines_offset = used - half_height;
+          loff.height -= plines_offset;
+          used = half_height;
+
+          topline = loff.lnum;
+          topfill = loff.fill;
+          skipcol = skipcol_from_plines(curwin, plines_offset);
+          set_skipcol = true;
+        }
+        break;
+      }
+      topline = loff.lnum;
+      topfill = loff.fill;
+      continue;
+    }
+
+    // If not using smoothscroll, we have to iteratively find how many
+    // lines to scroll down to roughly fit the cursor.
+    // This may not be right in the middle if the lines' physical height >
+    // 1 (e.g. 'wrap' is on).
+
```

Comment:
Duplicate comment
```suggestion
```

Reply:
Huh I guess it was added by accident when porting a different patch?

Reply:
No, I added it intentionally because the position of the linebreak changed.

Reply:
Do you want me to change the linebreak to stay compatible with vim? I may have changed the linebreak some other comments as well. I thought we allowed/preferred longer lines.

I guess the point is being able to `patch` without conflicts? I did everything by hand because most of it doesn't apply cleanly anyways because of ifdefs/formatting...

Reply:
Hmm, I do think having a line break after `>` sign looks a bit strange.

Reply:
Sure

Reply:
> No, I added it intentionally because the position of the linebreak changed. The existing comment has the latest linebreak position.

I meant that this comment was originally added in a smoothscroll patch, so it was unexpected to already have in our code. Hence why I missed it and it ended up duplicated.